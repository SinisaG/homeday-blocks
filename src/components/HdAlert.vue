<template>
  <div
    class="alert"
    :class="{
      [`alert--${type}`]: true,
      'alert--no-icon': noIcon,
    }"
  >
    <img
      v-if="!noIcon"
      :src="iconSrc"
      class="alert__icon"
    >
    <p
      class="alert__message"
    >
      <slot />
    </p>
  </div>
</template>

<script>
import successIcon from '@/assets/icons/ic_success.svg';
import infoIcon from '@/assets/icons/ic_info.svg';
import warningIcon from '@/assets/icons/ic_warning.svg';
import errorIcon from '@/assets/icons/ic_error.svg';

export const TYPES = {
  success: {
    icon: successIcon,
  },
  info: {
    icon: infoIcon,
  },
  warning: {
    icon: warningIcon,
  },
  error: {
    icon: errorIcon,
  },
};

export default {
  name: 'HdAlert',
  props: {
    type: {
      type: String,
      default: 'info',
      validator(type) {
        return TYPES[type] !== undefined;
      },
    },
    icon: {
      type: String,
      default: '',
    },
    noIcon: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    iconSrc() {
      return this.icon || TYPES[this.type].icon;
    },
  },
};
</script>

<style lang="scss" scoped>
.alert {
  display: flex;
  padding: $inset-s;
  @include font('text-small');
  border-width: 1px;
  border-style: solid;
  border-radius: 3px;

  &--success {
    border-color: $mantis;
    color: $mantis;
    background-color: rgba($mantis, .07);
  }

  &--info {
    border-color: $heather;
    color: $nevada;
  }

  &--warning {
    border-color: $crusta;
    color: $crusta;
    background-color: rgba($crusta, .07);
  }

  &--error {
    border-color: $torch-red;
    color: $torch-red;
    background-color: rgba($torch-red, .07);
  }

  &--no-icon {
    padding: $stack-s $inline-m;
  }

  &__icon {
    width: 24px;
    align-self: flex-start;
    margin-top: $stack-xxs;
    margin-right: $inline-s;
  }
}
</style>
