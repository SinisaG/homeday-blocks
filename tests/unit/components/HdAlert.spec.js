import { shallowMount } from '@vue/test-utils';
import HdAlert from '@/components/HdAlert.vue';

const MESSAGE_SELECTOR = '.alert__message';
const ICON_SELECTOR = '.alert__icon';

const CONTENT = '<b>Default slot</b>';

describe('HdAlert', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(HdAlert, {
      slots: {
        default: CONTENT,
      },
    });
  });

  test('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('renders the slot correctly', () => {
    expect(wrapper.find(MESSAGE_SELECTOR).element.innerHTML).toBe(CONTENT);
  });

  test('renders the right type', () => {
    // It doesn't make sense to go through all the possible types (we may have many)
    // they share the same logic, so if one works the rest should work as well
    const TEST_TYPE = 'warning';

    wrapper.setProps({
      type: TEST_TYPE,
    });

    expect(wrapper.classes()).toContain(`alert--${TEST_TYPE}`);
  });

  test('renders a custom icon', () => {
    const ICON_URL = 'https://placeholder.pics/svg/24x24';
    wrapper.setProps({
      icon: ICON_URL,
    });

    expect(wrapper.find(ICON_SELECTOR).attributes().src).toBe(ICON_URL);
  });

  test('doesn\'t render an icon if the prop `no-icon` is true', () => {
    wrapper.setProps({
      noIcon: true,
    });

    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);
  });
});
