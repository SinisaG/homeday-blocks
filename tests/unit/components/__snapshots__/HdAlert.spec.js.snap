// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdAlert renders component 1`] = `
<div class="alert alert--info"><img src="" class="alert__icon">
  <p class="alert__message"><b>Default slot</b></p>
</div>
`;
